tools/depcheck/pkg/cmd/trace/trace.go

```diff
@@ -0,0 +1,428 @@
+	}
+
+	fmt.Printf("Analyzing a total of %v packages\n", len(g.Nodes()))
+	fmt.Println()
+
+	// determine existing orphans and roots
+	keep := []graph.Node{}
+	roots := []graph.Node{}
+	for _, n := range g.Nodes() {
+		node, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		to := g.To(n)
+		from := g.From(n)
+
+		// found an existing orphaned node to exclude
+		if len(to) == 0 && len(from) == 0 {
+			continue
+		}
+
+		// found a root
+		if len(from) > 0 && len(to) == 0 {
+			roots = append(roots, n)
+			continue
+		}
+
+		// skip package we are analyzing
+		if node.UniqueName == o.Analyze {
+			continue
+		}
+
+		keep = append(keep, n)
+	}
+
+	yoursGraph, err := copyGraph(g)
+	if err != nil {
+		return err
+	}
+	yours, err := printYours(yoursGraph, roots, keep, you)
+	if err != nil {
+		return err
+	}
+	fmt.Println()
+
+	// no need to copy the graph a second time - we can mutate it
+	// since it won't be used anymore
+	mine, err := printMine(g, roots, keep, you)
+	if err != nil {
+		return err
+	}
+
+	ours := []graph.Node{}
+	yoursOrMine := append(yours, mine...)
+	for _, n := range keep {
+		isYoursOrMine := false
+		for _, yom := range yoursOrMine {
+			if n.ID() == yom.ID() {
+				isYoursOrMine = true
+				break
+			}
+		}
+		if !isYoursOrMine {
+			ours = append(ours, n)
+		}
+	}
+	fmt.Println()
+
+	fmt.Printf("\"Ours\": %v packages shared between the main repo and %q\n", len(ours), you.UniqueName)
+	for _, n := range ours {
+		pkg, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		fmt.Printf("    - %s\n", pkg.UniqueName)
+	}
+
+	return nil
+}
+
+func printYours(g *depgraph.MutableDirectedGraph, roots, nodes []graph.Node, you *depgraph.Node) ([]graph.Node, error) {
+	g.RemoveNode(you)
+
+	// find nodes not reachable from our given roots
+	yours := findOrphans(g, roots, nodes)
+
+	fmt.Printf("\"Yours\": %v packages exclusive to %q\n", len(yours), you.UniqueName)
+	for _, n := range yours {
+		pkg, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		fmt.Printf("    - %s\n", pkg.UniqueName)
+	}
+
+	return yours, nil
+}
+
+func printMine(g *depgraph.MutableDirectedGraph, roots, nodes []graph.Node, you *depgraph.Node) ([]graph.Node, error) {
+	for _, r := range roots {
+		g.RemoveNode(r)
+	}
+
+	// find nodes not reachable from "you"
+	mine := findOrphans(g, []graph.Node{you}, nodes)
+	fmt.Printf("\"Mine\": %v packages exclusive to the main repo\n", len(mine))
+	for _, n := range mine {
+		pkg, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		fmt.Printf("    - %s\n", pkg.UniqueName)
+	}
+
+	return mine, nil
+}
+
+// findOrphans receives a set of root nodes and a set of child nodes and
+// returns the set of child nodes not reachable from any root node.
+func findOrphans(g graph.Directed, roots []graph.Node, nodes []graph.Node) []graph.Node {
+	orphans := []graph.Node{}
+
+	for _, n := range nodes {
+		isOrphan := true
+		for _, root := range roots {
+			if hasPathFromTo(g, root, n) {
```

Comment:
This doesn't look right. If you started your graph without orphans, then any orphans you're finding later on the result of the removing the "from" on an inbound link.